Prevent analytics import orders action from being scheduled when analytics is disabled

ianlin opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #56616.
Closes #54791.
Closes #55123.

This PR prevents the `wc-admin_import_orders` analytics action from being scheduled when the analytics feature is disabled.

The issue was introduced in [#52480](https://github.com/woocommerce/woocommerce/pull/52480), which aimed to stop admin feature classes from loading on the frontend. However, it also [checks in get_feature_class()](https://github.com/woocommerce/woocommerce/pull/52480/files#diff-445f2f2cde958786c92a21d75fbed58a94a5594efc1f2056843fb4c6a9ccb3a7R114-R117), which is called by [Features::is_enabled](https://github.com/woocommerce/woocommerce/blob/59ae21fd3bc070153eae5d095cb76ba022996dc1/plugins/woocommerce/src/Admin/Features/Features.php#L206) -> [Features::get_available_features()](https://github.com/woocommerce/woocommerce/blob/59ae21fd3bc070153eae5d095cb76ba022996dc1/plugins/woocommerce/src/Admin/Features/Features.php#L176). If the feature class doesn't exist, it defaults to returning all features, including analytics, even when analytics is disabled.

Since `get_feature_class()` returns the class as a string, it's fine to remove the check while still fulfilling the purpose of #52480.

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Use WC >= 9.6
2. Disable WC Analytics feature in `WooCommerce > Settings > Advanced > Features`
3. Purchase some products so an order is created
4. Go to `WooCommerce > Status > Scheduled Actions`, see the action `wc-admin_import_orders` is not scheduled
5. Verifying the unit test is passed by running ~`pnpm --filter="@woocommerce/plugin-woocommerce" test:php:env --testsuite=wc-phpunit-legacy -- --filter 'WC_Admin_Tests_Analytics_Feature'`~ `pnpm --filter="@woocommerce/plugin-woocommerce" test:php:env --testsuite=wc-phpunit-main -- --filter 'FeatureEnabledTest'`
6. Revert the c1e0444005f88b6a06b10419602675e9055ba948 locally and run the test, see the test is failed

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Fix - Prevent analytics import orders action from being scheduled when analytics is disabled

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ianlin committed:
Still get feature class as the string when not in admin

ianlin committed:
Add a test for analytics feature enabled

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

ianlin commented:
Hey @chihsuan, tagging you as a reviewer since you're the author of #52480 and I'd like to make sure I'm not missing something. 🙏

chihsuan reviewed (commented):
Thanks for the investigation and the fix! @ianlin Apologize for the problems caused by #52480

> Since get_feature_class() returns the class as a string, it's fine to remove the check while still fulfilling the purpose of #52480.

I think that eliminating the check in `get_feature_class()` will result in the feature classes being autoloaded when `get_feature_class()` is called. Because `class_exists` will trigger the autoloader and load the class. It doesn't initialize the class though.

Do you think there is an alternative solution to address this issue?

chihsuan commented on plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/features/analytics-feature.php:
```diff
@@ -0,0 +1,28 @@
+/**
+ * class WC_Admin_Tests_Analytics_Feature.
+ */
+class WC_Admin_Tests_Analytics_Feature extends WC_Unit_Test_Case {
```
We should move the tests to `woocommerce/plugins/woocommerce/tests/php/src/Admin/Features/`.

ianlin commented:
Thanks for reviewing and the suggestion @chihsuan.

> Because class_exists will trigger the autoloader and load the class. It doesn't initialize the class though.

Thanks for the heads up, and you're right that `class_exists` will auto load the classes if it's not loaded by default if we don't specify the 2nd parameter as false. However, I was following the testing instruction in #52840:

> 9. Add the following code to `Features.php` before [line 56](https://github.com/woocommerce/woocommerce/blob/eb4cfbcc4e5b10780d72bd7097e67b40c9d37afa/plugins/woocommerce/src/Admin/Features/Features.php#L55):
>
> ```php
> spl_autoload_register(
>	function( $classname ) {
>		error_log( print_r( $classname, true ) );
>	}
> );
>```
>
> 10. Add `define( 'DISABLE_WP_CRON', true );` to `wp-config.php` to prevent wp-cron from loading admin features by itself
> 11. Go to frontend in incognito mode and confirm no errors
> 12. Confirm `wp-content/debug.log` doesn't log any `Automattic\WooCommerce\Admin\Features\` classes

And I didn't see the Admin Feature classes were being loaded in the `debug.log`, not sure if I miss something?

ianlin committed:
Move unit test to the right place

ianlin committed:
Pass should_autoload_class to class_exists method

ianlin commented:
Either way, I added a commit to pass the 2nd parameter to `class_exists()` method so it's explicitly not loading the class while checking its existence. I also moved the test to the correct folder as you suggested, it's ready for another round of review.

chihsuan reviewed (approved):
> And I didn't see the Admin Feature classes were being loaded in the debug.log, not sure if I miss something?

I couldn't see the logs, either. I guess something might have changed.

> Either way, I added a commit to pass the 2nd parameter to class_exists() method so it's explicitly not loading the class while checking its existence. I also moved the test to the correct folder as you suggested, it's ready for another round of review.

Thanks for addressing the feedback, the changes look good to me! Tested well. 👍 

___

#### Test environment

<details>
- WordPress: 6.7.2
- Theme: Twenty Twenty-Five Version: 1.1
- WooCommerce: Development version with fix/56616-54791-55123-analytics-import-orders-action-scheduled-when-analytics-is-disabled branch
- PHP: 8.3
- Web Server: Nginx
- Browser: Brave
- OS: macOS Sequoia 15.3.1
</details>

#### Review checklist

| Parameter | Result |
|-----------|---------|
| Approach | ✅ Good - Addresses root cause by fixing class autoloading behavior |
| Implementation | ✅ Clean and focused change |
| Automated tests | ✅ Tests moved to correct location and cover core functionality |
| Naming | ✅ Clear and consistent |
| Typing | ✅ Appropriate use of types |
| Code comments | ✅ Sufficient |
| Changelog | ✅ Clear and descriptive |
| Documentation | ✅ Sufficient inline documentation |
| PR description | ✅ Well explained with context |
| Testing instructions | ✅ Clear and comprehensive |
| Manual testing | ✅ Works as expected |

#### Scenarios tested

| Scenario | Result | Screenshot |
|----------|---------|------------|
| Create new order with analytics disabled | ✅ No import action scheduled | |
| Create new order with analytics enabled | ✅ Import action scheduled correctly | <img width="1021" alt="Screenshot 2025-04-01 at 17 48 24" src="https://github.com/user-attachments/assets/1c5cb4b8-3280-44db-a22e-d546d4115a5b" /> |

chihsuan commented:
@ianlin  This isn't a blocker for the merge, but I wanted to check if we need to initialize these classes on the front end. Would you happen to have any insights on their necessity there?

https://github.com/woocommerce/woocommerce/blob/fec0008b32ec0895bb111205b8034fb89e2bc444/plugins/woocommerce/src/Internal/Admin/FeaturePlugin.php#L174-L181

I noticed that other admin classes are being initialized in the front end as well. Some of them, such as onboarding classes, are only required in the admin context. I think there is a chance to optimize the performance.

ianlin commented:
Thanks for the detailed review @chihsuan!

> I wanted to check if we need to initialize these classes on the front end. Would you happen to have any insights on their necessity there?
> https://github.com/woocommerce/woocommerce/blob/fec0008b32ec0895bb111205b8034fb89e2bc444/plugins/woocommerce/src/Internal/Admin/FeaturePlugin.php#L174-L181

The `ReportsSync` class is required on the frontend when the Analytics feature is enabled (i.e. when the `woocommerce_analytics_enabled` option is set to true via `WooCommerce > Settings > Advanced > Features`). It registers actions like `wc-admin_import_orders`, so when an order is created or updated, the action gets scheduled to import order data into the `wc_order_stats` table for analytics.

I'm not so sure but I think `ReportExport` class might not be required on the frontend, it would require some testing to understand.

> I noticed that other admin classes are being initialized in the front end as well. Some of them, such as onboarding classes, are only required in the admin context. I think there is a chance to optimize the performance.

Indeed, there are still room for performance improvement, but I think it's not an easy task to determine which is not needed. I can create an issue as a follow up. Thanks for pointing this out!

ianlin committed:
Merge branch 'trunk' into fix/56616-54791-55123-analytics-import-orders-action-scheduled-when-analytics-is-disabled

kaushikasomaiya commented:
@ianlin @chihsuan I cannot replicate this issue on my end but a merchant is seeing it - assume the PR takes care of it but I am curious how can we replicate this?

ianlin commented:
@kaushikasomaiya Did you follow the steps to reproduce in #56616? Those steps are sufficient enough to replicate the issue I think.

kaushikasomaiya commented:
@ianlin  Yes those steps did not help replicate - in merchant's case one thing I am sure is possibly_schedule_import() was getting hooked despite Analytics disabled from settings while in my site, possibly_schedule_import was not being hooked.

pawelkmpt commented:
It'd be nice if it was merged also to previous WooCommerce versions. I'm running 9.7.1 and this is spamming my log heavily.

pawelkmpt commented:
> @pawelkmpt Sorry for causing you any trouble! We don't really support patching older versions. Any reasons that you want to keep using 9.7.1 instead of updating to the latest version?

Updating to the latest version immediately is usually risky, so I prefer to wait couple of weeks until given release is stable.

I made a workaround fix with attaching empty callback to the action to prevent spamming. Will remove it once I'm on 9.9.
